subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc

```diff
@@ -0,0 +1,95 @@
+
+[NOTE]
+====
+The third-party link:https://github.com/nebula-plugins/gradle-lint-plugin[Gradle lint plugin] helps with enforcing a desired code style in a build script if you are looking for appropriate link:https://en.wikipedia.org/wiki/Lint_(software)[linting] automation.
+====
+
+[[sec:declare_task_group_and_description]]
+=== Declare the group and description for tasks
+
+Even new users to a build should to be able to find crucial information quickly and effortlessly. In Gradle you can declare a api:org.gradle.api.Task#setGroup(java.lang.String)[group] and a api:org.gradle.api.Task#setDescription(java.lang.String)[description] for any task of the build. The <<sec:listing_tasks,tasks report>> uses the assigned values to organize and render the task for easy discoverability. Assigning a group and description is most helpful for lifecycle tasks.
+
+The example task `generateDocs` generates documentation for a project in the form of HTML pages. The task should be organized underneath the bucket `Documentation`. The description should express its intent.
+
+++++
+<sample id="taskGroupDescription" dir="userguide/bestPractices/taskGroupDescription" title="A task declaring the group and description">
+    <sourcefile file="build.gradle"/>
+</sample>
+++++
+
+The output of the tasks report reflects the assigned values.
+
+----
+$ gradle tasks
+
+> Task :tasks
+
+...
+
+Documentation tasks
+-------------------
+generateDocs - Generates the HTML documentation for this project.
+
+...
+----
+
+[[sec:avoid_use_of_imperative_logic]]
+=== Avoid the use of imperative logic
+
+The Gradle runtime does not enforce a specific style for build logic. For that very reason, it's easy to end up with a build script that mixes declarative DSL elements with imperative, procedural code. Let's talk about some concrete examples.
+
+* _Declarative code:_ Built-in, language-agnostic DSL elements (e.g. api:org.gradle.api.Project#dependencies(groovy.lang.Closure)[] or api:org.gradle.api.Project#repositories(groovy.lang.Closure)[]) or DSLs exposed by plugins
+* _Imperative code:_ Conditional logic or very complex task action implementations
+
+The end goal of every build script should be to only contain declarative language elements which makes the code easier to understand and maintain. Imperative logic should live in binary plugins and which in turn is applied to the build script. As a side product, you automatically enable your team to link:https://guides.gradle.org/designing-gradle-plugins/#reusable_logic_should_be_written_as_binary_plugin[reuse the plugin logic in other projects] if you publish the artifact to a binary repository.
```

Comment:
I'd say "live in `buildSrc` or binary plugins" and perhaps link to appropriate sections of "Organizing Build Scripts" or whatever chapter it'll be.

Reply:
The important point I am trying to make here is that you actually need to publish your artifact to be able to reuse it in other project. `buildSrc` and binary plugin won't do per se. The dependency actually needs to be available from another project (that likely lives in a different repo).

Reply:
Ok